File: app/src/main/java/com/june/mvpkj/j/base/BaseActivity.java
package com.june.mvpkj.j.base;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

/**
 * @CreateTime: 2022/1/14
 * @author: jjzhu
 */
public abstract class BaseActivity<P extends BasePresenter> extends AppCompatActivity {

    protected P presenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(layout());
        presenter = createPresenter();
        init();
    }

    protected abstract P createPresenter();

    protected abstract View layout();

    protected abstract void init();

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }
}

File: app/src/main/java/com/june/mvpkj/k/base/BaseActivity.kt
package com.june.mvpkj.k.base

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity

/**
 *@CreateTime: 2022/1/14
 *@author: jjzhu
 */
abstract class BaseActivity<V : BaseView ,P : BasePresenter<V>> : AppCompatActivity() {

    var presenter : P? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout())
        presenter = createPresenter()
        init()
    }

    abstract fun createPresenter() : P

    abstract fun layout(): View

    abstract fun init()

    override fun onDestroy() {
        super.onDestroy()
        presenter?.detachView()
    }
}